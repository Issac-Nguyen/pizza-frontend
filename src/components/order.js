import React from 'react'
import {connect} from 'react-redux'
import {Row, Col, Collapse, List, Divider, Alert} from 'antd'
import * as _ from 'lodash'
import {format} from 'date-fns'

const { Panel } = Collapse;

export const Order = (props) => {
    let orders = _.groupBy(props.order, o => (format(new Date(o.created_at), 'yyyy/MM/dd')));

    const renderOrderItem = (orderItems) => {
        return (
            <List
                header={<div>Booked pizzas</div>}
                footer={<div></div>}
                bordered
                dataSource={orderItems}
                renderItem={item => (
                    <Row className="order-order-item">
                        <Col span={6} className="order-order-img"><img src={item.url} /></Col>
                        <Col span={6}>{item.name}</Col>
                        <Col span={6}>${item.price}</Col>
                    </Row>
                )}
    />
        )
    }

    const renderNest = (id, order) => {
        const header = <Row>
            <Col span={5}>
                {order.customer_name}
            </Col>
            <Col span={5}>
                {order.customer_address}
            </Col>
            <Col span={5}>
                {order.customer_phone}
            </Col>
            <Col span={5}>
                Delivery: {order.schedule}
            </Col>
            <Col span={4}>
            {`Total: $${order.total} (Eur ${(order.total * props.rate).toFixed(2)})`}
            </Col>
        </Row>

        return (
            <Collapse key={id}>
                <Panel header={header} key="1" className="order-items-panel">
                    {renderOrderItem(order.order_items)}
                </Panel>
            </Collapse>
        )
    }

    let arrayPanel  = [];
    _.forEach(orders, (vl, k) => {
        arrayPanel.push(<Panel header={k} key={k}>
            <Collapse>
                {vl.map((e, index) => renderNest(index, e))}
            </Collapse>
        </Panel>)
    })
        
    return (
        <Row>
            <Col span={24}>
            <Divider>Your orders:</Divider>
            
                {arrayPanel.length > 0 && (
                    <Collapse>
                    {arrayPanel}
                </Collapse>
                )}
               {arrayPanel.length == 0 && (
                <Row justify="center">
                    <Alert message="You don't have any order today. Please book our pizzas" type="error" />
                </Row>
               )}
                </Col>
               </Row>
    )
}

const mapStateToProps = state => ({
    login: state.login,
    order: state.order,
    rate: state.rate
})
export default connect(mapStateToProps)(Order);