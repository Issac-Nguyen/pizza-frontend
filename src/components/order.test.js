import React from 'react'
import {render, cleanup} from '@testing-library/react'
import {Order} from './order'

Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });

describe('Order', () => {
    afterEach(cleanup);
 
    it('should take a snapshot with order has no length', () => {
       const { asFragment } = render(<Order order={[]}/>)
       
       expect(asFragment(<Order order={[]}/>)).toMatchSnapshot()
    })

    it('should take a snapshot with order has length', () => {
        const order = {created_at: 1588061743099, customer_address: "Ô 18 Lô 903 Hòn Rớ, Phước Đồng",
            customer_name: "Nguyen Huu Phat",
            customer_phone: "0777955777",
            email: "",
            fee: 4.66,
            order_items: [
                {
                currency: "$",
                id: 4,
                name: "pizza 4",
                pizza_id: 4,
                price: 10,
                url: "https://image.shutterstock.com/image-photo/concept-promotional-flyer-poster-restaurants-260nw-1060535249.jpg"
                },
                {
                    currency: "$",
                    id: 6,
                    name: "pizza 4",
                    pizza_id: 4,
                    price: 10,
                    url: "https://image.shutterstock.com/image-photo/concept-promotional-flyer-poster-restaurants-260nw-1060535249.jpg"
                    }
            ],
            schedule: true,
            sendimmediate: true,
            total: "24.66"};

        const { asFragment } = render(<Order order={[order]}/>)
        
        expect(asFragment(<Order order={order}/>)).toMatchSnapshot()
     })

     it('should show no order if no order in props', () => {
        const { getByText } = render(<Order order={[]}/>)
        
        expect(getByText(/you don't have any order today/i)).toBeInTheDocument()
     })

     it('should show your orders with order has length', () => {
        const order = {created_at: 1588061743099, customer_address: "Ô 18 Lô 903 Hòn Rớ, Phước Đồng",
            customer_name: "Nguyen Huu Phat",
            customer_phone: "0777955777",
            email: "",
            fee: 4.66,
            order_items: [
                {
                currency: "$",
                id: 4,
                name: "pizza 4",
                pizza_id: 4,
                price: 10,
                url: "https://image.shutterstock.com/image-photo/concept-promotional-flyer-poster-restaurants-260nw-1060535249.jpg"
                },
                {
                    currency: "$",
                    id: 6,
                    name: "pizza 4",
                    pizza_id: 4,
                    price: 10,
                    url: "https://image.shutterstock.com/image-photo/concept-promotional-flyer-poster-restaurants-260nw-1060535249.jpg"
                    }
            ],
            schedule: true,
            sendimmediate: true,
            total: "24.66"};

        const { getByText } = render(<Order order={[order]}/>)
        
        expect(getByText(/your orders/i)).toBeInTheDocument()
     })

     it('should show collapse with order has length', () => {
        const order = {created_at: 1588061743099, customer_address: "Ô 18 Lô 903 Hòn Rớ, Phước Đồng",
            customer_name: "Nguyen Huu Phat",
            customer_phone: "0777955777",
            email: "",
            fee: 4.66,
            order_items: [
                {
                currency: "$",
                id: 4,
                name: "pizza 4",
                pizza_id: 4,
                price: 10,
                url: "https://image.shutterstock.com/image-photo/concept-promotional-flyer-poster-restaurants-260nw-1060535249.jpg"
                },
                {
                    currency: "$",
                    id: 6,
                    name: "pizza 4",
                    pizza_id: 4,
                    price: 10,
                    url: "https://image.shutterstock.com/image-photo/concept-promotional-flyer-poster-restaurants-260nw-1060535249.jpg"
                    }
            ],
            schedule: true,
            sendimmediate: true,
            total: "24.66"};

        const { getByRole } = render(<Order order={[order]}/>)
        
        expect(getByRole('button')).toBeInTheDocument()
        expect(getByRole('button')).toHaveAttribute('aria-expanded')
     })
})
