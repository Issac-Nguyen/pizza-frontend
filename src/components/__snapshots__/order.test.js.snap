// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order should take a snapshot with order has length 1`] = `
<DocumentFragment>
  <div
    class="ant-row"
  >
    <div
      class="ant-col ant-col-24"
    >
      <div
        class="ant-divider ant-divider-horizontal ant-divider-with-text-center"
        role="separator"
      >
        <span
          class="ant-divider-inner-text"
        >
          Your orders:
        </span>
      </div>
      <div
        class="ant-collapse ant-collapse-icon-position-left"
      >
        <div
          class="ant-collapse-item"
        >
          <div
            aria-expanded="false"
            class="ant-collapse-header"
            role="button"
            tabindex="0"
          >
            <span
              aria-label="right"
              class="anticon anticon-right ant-collapse-arrow"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                data-icon="right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                />
              </svg>
            </span>
            2020/04/28
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Order should take a snapshot with order has no length 1`] = `
<DocumentFragment>
  <div
    class="ant-row"
  >
    <div
      class="ant-col ant-col-24"
    >
      <div
        class="ant-divider ant-divider-horizontal ant-divider-with-text-center"
        role="separator"
      >
        <span
          class="ant-divider-inner-text"
        >
          Your orders:
        </span>
      </div>
      <div
        class="ant-row ant-row-center"
      >
        <div
          class="ant-alert ant-alert-error ant-alert-no-icon"
          data-show="true"
        >
          <span
            class="ant-alert-message"
          >
            You don't have any order today. Please book our pizzas
          </span>
          <span
            class="ant-alert-description"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
